### PR TITLE
Identify the correct line where mixed whitespace is used in a file

### PR DESCRIPTION
**Pull request by Amazinite**
**Bugfix:** This PR addresses the bug highlighted in issue #6410, as well as the bugs highlighted in https://github.com/endless-sky/endless-sky/pull/6411#issuecomment-996186516.

## Fix Details
When the game parses a data file, it will warn if it finds that the file uses mixed whitespace indentation. That is, if the file uses tabs but inside one of the lines it switches to using spaces, that line will be warned about for having mixed whitespace. Similarly, if a file is using spaces for indentation but switches to using tabs, that will also be warned about. The line where the mixed whitespace is used is identified by a PrintTrace.

There are two different warnings for mixed whitespace:
* "Mixed whitespace usage in file": Only appears once per file. Triggered if a line starts with a different indentation than what was used prior. That is, a tabs indentation file having a line that starts with a space, or vice versa.
* "Mixed whitespace usage in line": Appears as many times for as many lines trigger this warning. Triggered if a line contains a different indentation than what was used prior but that different indentation isn't at the start of the line. For example, a tabs indentation file having a line with indentation "tab, space".

There are actually some bugs with this system right now, though:
1. "Mixed whitespace usage in file" doesn't work as one would expect. The way that it works is that if at any point a line starts with a space, a fileIsSpaces boolean is set to true. Then if at any point after fileIsSpaces is set to true a tab is observed, that's when this error is called. This means that if a file starts by using tabs but then switches over to spaces, no error is thrown.
2. The PrintTrace call that writes the mixed whitespace error is called from the DataNode at the top of the stack just as the error is found. The problem with this is that the stack only contains DataNodes for lines that have been fully parsed, but the error is written while a line is parsing, meaning that the line that the PrintTrace points to is actually the line above where the error occurred.
3. As pointed out in #6410, the current system doesn't properly point to comments with mixed whitespace. This is because comments are never turned into DataNodes, and therefore a PrintTrace can never be called that points directly to them.

This PR does the following things:
* Collapses the "in file" and "in line" messages to a single "Mixed whitespace usage at line" error. It didn't make much sense to me to have two separate errors just for whether the mixed whitespace is at the start of the line or past the start of the line. This error also isn't capped for a given file, so it will appear as many times as there are whitespace errors. 
   * My reason for this was so that one doesn't need to keep fixing the error and relaunching the game one line at a time. I will note though that all it would take for every single line in a file to throw an error now would be to add a space to the first line of a file, which could cause a load of errors to occur when really the error is with a single line. So what I could do is cap the number of errors so that it points out the exact line that the error is on for say the first 5 lines, then says "And X other lines" at the end.
* DataFile now identifies files as either tabs files or spaces files, not just as spaces files. The first indentation observed is what the file is marked as, and when the other indentation type is seen is when an error is thrown instead of only throwing an error if a file switches from spaces to tabs and taking the first instance of spaces being used as marking the file as a spaces file.
* PrintTrace is now only called from the top of the stack after a line is finished parsing. This way we know that we're calling PrintTrace from the line that the error was encountered on.
* Comments with mixed whitespace are now identified by a line number instead of calling a PrintTrace.
* Unrelated to the bug fixes (naughty, I know), but I also moved the empty line for PrintTrace to the end of the message instead of the beginning because it irked me. By writing the empty line between PrintTraces at the start, the error log always started with an empty line, and the last PrintTrace had no empty line between it and any warnings that were later written to the error log.

## Testing Done
Added mixed whitespace to lines in category.txt, including on the comment lines. Observed how the warnings are now for the correct line that the mixed whitespace was on, and every mixed whitespace error is now logged instead of only logging certain types of errors once and other types any number of times.

## Save File
N/A

### CODE REVIEW

**Review by tehhowch — CHANGES_REQUESTED** [review]
Needs unit tests for the expected behavior.

File: source/DataFile.cpp [hunk 1]
```diff
@@ -112,27 +112,22 @@ void DataFile::LoadData(const string &data)
 		size_t tokenPos = pos;
 		char32_t c = Utf8::DecodeCodePoint(data, pos);
 		
-		// Find the first non-white character in this line.
-		bool isSpaces = false;
+		bool mixedIndentation = false;
 		int white = 0;
+		// Find the first non-white character in this line.
```
[the remark this answers is not in the document]
Comment (a-random-lemurian) [reply source/DataFile.cpp:117]: `non-white` can be:

- `invisChars`
- `indentChars`
- `spaces`
- `tabs`

And I've ran out of ideas.

**Comment by Amazinite** [comment]
> Needs unit tests for the expected behavior.

Can someone else do that?

File: source/DataFile.cpp [hunk 2]
```diff
@@ -112,27 +112,22 @@ void DataFile::LoadData(const string &data)
 		size_t tokenPos = pos;
 		char32_t c = Utf8::DecodeCodePoint(data, pos);
 
-		// Find the first non-white character in this line.
-		bool isSpaces = false;
+		bool mixedIndentation = false;
 		int white = 0;
+		// Find the first non-white character in this line.
```
Comment (tehhowch) [inline source/DataFile.cpp:117]:
```suggestion
 		int separators = 0;
		// Find the first tokenizable character in this line (i.e. neither space nor tab).
```

File: source/DataFile.cpp [hunk 3]
```diff
@@ -101,8 +101,8 @@ void DataFile::LoadData(const string &data)
 	// new node added at the next deeper indentation level.
 	vector<DataNode *> stack(1, &root);
 	vector<int> whiteStack(1, -1);
```
Comment (tehhowch) [inline source/DataFile.cpp:103]:
```suggestion
 	vector<int> separatorStack(1, -1);
```

**Review by tehhowch — COMMENTED** [review]
Last bits

**Pull request merged**